### DISCUSSION TITLE
How to Unit Test App.addListener in Jest

### DISCUSSION THREAD

**Discussion by robmarshall** [ionic-team/capacitor#5477] [Q&A]
Hi,

I am trying to write Jest unit tests for events passed from App.

More specifically - how can I test if someFunction has fired.

```
import { App } from '@capacitor/app';
export const useSomeHook = () => {
    useEffect(() => {
        App.addListener('appStateChange', ({ isActive }) => {
            if (isActive) {
                someFunction();
            }
        });
    }, []);
};
```

The docs on testing seem limited, and I cannot see any issues on this. Am I missing a simple thing?

Thanks in advance!

**Comment by develmood** [comment]
In the first step, you have to mock the addListener function. The important point here is, you have to call the `listenerFunc` in your mock, because in your implementation you insert here your anonymous function `({ isActive }) => ...`, which gets not called if you don't call it in your mock. 

```
jest.mock('@capacitor/app', () => ({
   App: {
    addListener: (eventName: string, listenerFunc: StateChangeListener) => listenerFunc({ isActive: true }),
  }
});
```

Then you can define the assertion

```
it('should call someFunction()', () => {
  const spy = jest.spyOn(testUnit, 'someFunction');
  testUnit.useSomeHook();

  expect(spy).toHaveBeenCalled();
});
```

Edit: I never used React, so I don't know how to adjust the test to work with useEffect, but this is the basic structure to test function calls from external plugins.
